src/psyclone/psyir/tools/dependency_tools.py

```diff
@@ -273,102 +224,443 @@ def _is_loop_suitable_for_parallel(self, loop, only_nested_loops=True):
+        :rtype: list of sets of str
 
-        * if the array access does not depend on the loop variable, a
-          warning is added (e.g. for the variable `a` in `a(1,2) = b(i,j)`).
-        * if the array variable is accessed inconsistently, e.g.
-          `a(i,j) = a(j,i) + 1`.
+        '''
+        indices = []
+        for i in component_indices.iterate():
+            indx = component_indices[i]
+            index_vars = VariablesAccessInfo(indx)
+            unique_vars = set(str(sig) for sig in index_vars.keys())
+            unique_vars = unique_vars.intersection(set_of_loop_vars)
+            indices.append(unique_vars)
+        return indices
 
-        :param loop_variable: symbol of the variable that is parallelised.
-        :type loop_variable: :py:class:`psyclone.psyir.symbol.DataSymbol`
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def partition(comp_ind1, comp_ind2, loop_vars):
+        '''This function partitions the subscripts to the component indices
+        into disjunct sets. For example:
+        `a(i)` and `a(i+3)` results in one partition with the variable `i`,
+        `a(i)` and `a(j)` results in one partition with the variable `i`, `j`,
+        `a(i,j,k)` and `a(i,k,j)` results in two partitions,
+            one for subscript 0 (variable `i`),
+            one for subscripts 1 and 2n (variables `j` and `k`)
+
+        :param comp_ind1: component_indices of the first array access.
+        :type comp_ind1:  \
+            :py:class:`psyclone.core.component_indices.ComponentIndices`
+        :param comp_ind2: component_indices of the first array access.
+        :type comp_ind2:  \
+            :py:class:`psyclone.core.component_indices.ComponentIndices`
+
+        :return: partition information.
+        :rtype: list of 2-tuple of set-of-str and list of integer
+
+        '''
+        # Get the (string) name of all variables used in each subscript
+        # of the two accesses. E.g. `a(i,j+k)` --> [ {"i"}, {"j","k"}]
+        set_of_loop_vars = set(loop_vars)
+        indices_1 = DependencyTools.get_flat_indices(comp_ind1,
+                                                     set_of_loop_vars)
+        indices_2 = DependencyTools.get_flat_indices(comp_ind2,
+                                                     set_of_loop_vars)
+        # This list stores the partition information, which
+        # is a pair consisting of:
+        # - a set of all loop variables used in the subscript of
+        #   both accesses
+        # - list of all subscripts. Initially these lists contain
+        #   only one subscript, but they will be modified later
+        # Example: `a(i,j)` and `a(i,k)` -->
+        #          [ ({"i"}, [(0,0)]), ({"j","k"}, [(0,1)])]
+        partition_infos = []
+        for i, indx in enumerate(comp_ind1.iterate()):
+            partition_infos.append((indices_1[i].union(indices_2[i]), [indx]))
```

Comment:
Some Nemo files fail here with a IndexOutOfRange. I can not see with the debugger which Fortran expression produces it but I can dig deeper if you need.

For the time being could you add some test parametrizations in src/psyclone/tests/psyir/tools/dependency_tools_test.py::test_partitions where "comp_ind1" and "comp_ind2" have different number of components. As there isn't any.
I would also like to see cases where the loop_vars make an empty set in some components of one component_indices but not the other one e.g. "a(n)%b(n)", "a(i)%b(n)"